### PR TITLE
Add ECN support: ECN validation

### PR DESCRIPTION
**Pull request by csujedihy**
## Description

This PR implements [13.4.2. ](https://www.rfc-editor.org/rfc/rfc9000.html#section-13.4.2)[ECN Validation](https://www.rfc-editor.org/rfc/rfc9000.html#name-ecn-validation) in RFC9000. The ECN validation algorithm implemented in this PR is very similar to the sample algorithm given in RFC9000 A.4. The details can be found at the top of path.h.

This PR also adds a new global QUIC setting, EcnEnabled and a flag, EcnCapable, in connection stats.

This PR does not implement CC ECN reaction, which will be done in a separate PR.

## Testing

CI.

### CODE REVIEW

File: src/core/connection.h [hunk 1]
```diff
@@ -563,6 +564,18 @@ typedef struct QUIC_CONNECTION {
...
+    // ECT and CE counters for different packet number spaces.
+    //
+    uint64_t EcnEctCounters[QUIC_ENCRYPT_LEVEL_COUNT];
+    uint64_t EcnCeCounters[QUIC_ENCRYPT_LEVEL_COUNT]; // maps to ecn_ce_counters in RFC 9002.
+
+    //
+    // The total number of packets sent with each corresponding ECT codepoint in all encryption
+    // level.
+    //
+    uint64_t NumPacketsSentWithEct;
```
Comment (nibanks) [inline src/core/connection.h:577]: Can this be moved to a sub struct? I feel like it shouldn't be on the connection itself.

Comment (nibanks) [reply src/core/connection.h:577]: Should it be per-path?

File: src/core/sent_packet_metadata.h [hunk 2]
```diff
@@ -94,6 +94,7 @@ typedef struct QUIC_SEND_PACKET_FLAGS {
     //
     BOOLEAN IsAppLimited            : 1;
     BOOLEAN HasLastAckedPacketInfo  : 1;
+    BOOLEAN EcnEctSet               : 1;
```
Comment (nibanks) [inline src/core/sent_packet_metadata.h:97]: This makes us use another byte per packet. I wonder if there's a way to eliminate this need.

File: src/core/connection.h (hunk 1 above)
Comment (csujedihy) [reply src/core/connection.h:577]: The ECN count feedback isn't per path, right? If we make this per-path, it will be unsync'd with ECN counts from the peer.

File: src/test/lib/TestHelpers.h [hunk 3]
```diff
@@ -581,6 +581,26 @@ class DatapathHooks
     }
 };
 
+struct EcnModifyHelper : public DatapathHook
+{
+    CXPLAT_ECN_TYPE EcnType;
```
Comment (nibanks) [inline src/test/lib/TestHelpers.h:586]: Default?

File: src/core/sent_packet_metadata.h (hunk 2 above)
Comment (csujedihy) [reply src/core/sent_packet_metadata.h:97]: That's not possible I think. We need to know which packets are sent with ECT bit set.

Are you sure this will use another byte? I don't see clear effort for struct packing in QUIC_SENT_PACKET_METADATA.

File: src/test/lib/TestHelpers.h (hunk 3 above)
Comment (csujedihy) [reply src/test/lib/TestHelpers.h:586]: Oh right, it actually needs a IsSet flag that is default FALSE.

Comment (csujedihy) [reply src/test/lib/TestHelpers.h]: fixed

File: src/core/quicdef.h [hunk 4]
```diff
@@ -506,6 +506,11 @@ CXPLAT_STATIC_ASSERT(
 //
 #define QUIC_DEFAULT_GREASE_QUIC_BIT_ENABLED         FALSE
 
+//
+// The default value for enabling sender-side ECN support.
+//
+#define QUIC_DEFAULT_ECN_ENABLED         FALSE
```
Comment (nibanks) [inline src/core/quicdef.h:512]: nit: can you align the `FALSE` with the other values above?

File: src/core/path.h [hunk 5]
```diff
@@ -67,6 +101,21 @@ typedef struct QUIC_PATH {
...
 
+    //
+    // ECN validation state.
+    //
+    uint8_t EcnValidationState : 2;
+
+    //
+    // Indicates whether EcnTestingEndingTime is set.
+    //
+    BOOLEAN EcnTestingEndingTimeSet : 1;
```
Comment (nibanks) [inline src/core/path.h:112]: Can you just consider a value of `0` for `EcnTestingEndingTime` as unset?

File: src/core/sent_packet_metadata.h (hunk 2 above)
Comment (nibanks) [reply src/core/sent_packet_metadata.h:97]: I validated the size in the past. I can't claim 100% certainty now.

File: src/core/connection.h [hunk 6]
```diff
@@ -194,6 +194,7 @@ typedef union QUIC_CONNECTION_STATE {
         //
         BOOLEAN DisableVneTp : 1;
 #endif
+
```
Comment (nibanks) [inline src/core/connection.h:197]: nit: unnecessary change

File: src/core/loss_detection.c [hunk 7]
```diff
@@ -1496,6 +1498,69 @@ QuicLossDetectionProcessAckBlocks(
...
+            // Per RFC 9000, we validate ECN counts from received ACK frames
+            // when the largest acked packet number increases.
+            //
+            BOOLEAN EcnValidated = TRUE;
+            int64_t EctCeDeltaSum = 0;
+            if (Ecn != NULL) {
+                EctCeDeltaSum +=
+                    Ecn->CE_Count - Connection->EcnCeCounters[EncryptLevel];
+                EctCeDeltaSum +=
+                    Ecn->ECT_0_Count - Connection->EcnEctCounters[EncryptLevel];
```
Comment (nibanks) [inline src/core/loss_detection.c:1512]: Can we add some debug asserts to make sure the subtraction doesn't ever underflow?

File: src/core/loss_detection.c [hunk 8]
```diff
@@ -1496,6 +1498,69 @@ QuicLossDetectionProcessAckBlocks(
...
+                    EcnValidated = FALSE;
+                }
+
+                if (EcnValidated) {
+                    //
+                    // TODO: Notify CC of the ECN signal before we update the CE counts.
+                    //
+                    Connection->EcnCeCounters[EncryptLevel] = Ecn->CE_Count;
+                    Connection->EcnEctCounters[EncryptLevel] = Ecn->ECT_0_Count;
+                    if (Path->EcnValidationState <= ECN_VALIDATION_UNKNOWN) {
```
Comment (nibanks) [inline src/core/loss_detection.c:1531]: Is this check necessary? It's just excluding FAILED, right? We're already in a conditional that says !FAILED. Just add a debug assert instead?

File: src/core/loss_detection.c [hunk 9]
```diff
@@ -1740,12 +1805,15 @@ QuicLossDetectionProcessTimerOperation(
 
     if (OldestPacket != NULL &&
         CxPlatTimeDiff32(OldestPacket->SentTime, TimeNow) >=
-            MS_TO_US(Connection->Settings.DisconnectTimeoutMs)) {
+            MS_TO_US(Connection->Settings.DisconnectTimeoutMs) &&
+        (Path->EcnValidationState != ECN_VALIDATION_TESTING ||
+            !OldestPacket->Flags.EcnEctSet)) {
```
Comment (nibanks) [inline src/core/loss_detection.c:1810]: So, does this extend how long we wait for handshake failures because no response from the server?

File: src/core/packet_builder.c [hunk 10]
```diff
@@ -253,12 +253,36 @@ QuicPacketBuilderPrepare(
...
+                        uint32_t ThreePtosInUs =
+                            QuicLossDetectionComputeProbeTimeout(
+                                &Connection->LossDetection,
+                                &Connection->Paths[0],
+                                QUIC_CLOSE_PTO_COUNT);
+                        Builder->Path->EcnTestingEndingTime = TimeNow + ThreePtosInUs;
+                        Builder->Path->EcnTestingEndingTimeSet = TRUE;
+                    }
+                }
+            }
```
Comment (nibanks) [inline src/core/packet_builder.c:278]: I think this should either go (or be called from) at the beginning of `QuicSendFlush` or in `QuicPacketBuilderInitialize`.

Comment (nibanks) [reply src/core/packet_builder.c:278]: This is somewhat similar to `QuicMtuDiscoveryCheckSearchCompleteTimeout`.

File: src/core/packet_builder.c [hunk 11]
```diff
@@ -927,7 +951,7 @@ QuicPacketBuilderFinalize(
     Builder->Metadata->SentTime = CxPlatTimeUs32();
     Builder->Metadata->PacketLength =
         Builder->HeaderLength + PayloadLength;
-
+    Builder->Metadata->Flags.EcnEctSet = CxPlatSendDataEctSet(Builder->SendData);
```
Comment (nibanks) [inline src/core/packet_builder.c:954]: Instead of a platform abstraction, just save the EctSet flag on the builder.

File: src/inc/msquic.h [hunk 12]
```diff
@@ -473,7 +473,8 @@ typedef struct QUIC_STATISTICS_V2 {
     uint32_t ResumptionAttempted    : 1;
     uint32_t ResumptionSucceeded    : 1;
     uint32_t GreaseBitNegotiated    : 1;    // Set if we negotiated the GREASE bit.
-    uint32_t RESERVED               : 27;
+    uint32_t EcnCapable             : 1;
```
Comment (nibanks) [inline src/inc/msquic.h:476]: Should we expose some info about this feature in our statistics?

File: src/inc/quic_datapath.h [hunk 13]
```diff
@@ -586,6 +586,17 @@ CxPlatSendDataAlloc(
...
+// Returns whether ECT bit is set in the send context.
+//
+_IRQL_requires_max_(DISPATCH_LEVEL)
+_Success_(return != NULL)
+BOOLEAN
+CxPlatSendDataEctSet(
+    _In_ CXPLAT_SEND_DATA* SendData
+    );
+
+
```
Comment (nibanks) [inline src/inc/quic_datapath.h:599]:
```suggestion

```

File: src/test/bin/quic_gtest.cpp [hunk 14]
```diff
@@ -1826,6 +1826,17 @@ TEST(Misc, StreamAbortConnFlowControl) {
     }
 }
 
+#ifdef QUIC_TEST_DATAPATH_HOOKS_ENABLED
+TEST(Misc, Ecn) {
+    TestLogger Logger("Ecn");
+    if (TestingKernelMode) {
+        ASSERT_TRUE(DriverClient.Run(IOCTL_QUIC_RUN_ECN));
+    } else {
+        QuicTestEcn();
```
Comment (nibanks) [inline src/test/bin/quic_gtest.cpp:1835]: Should we have different IPv4 and IPv6 tests?

File: src/test/lib/DataTest.cpp [hunk 15]
```diff
@@ -2161,6 +2161,194 @@ QuicTestAbortReceive(
...
+    CxPlatEvent ServerStreamRecv;
+    CxPlatEvent ServerStreamShutdown;
+    MsQuicStream* ServerStream {nullptr};
+    bool ServerStreamHasShutdown {false};
+
+    static QUIC_STATUS StreamCallback(_In_ MsQuicStream* Stream, _In_opt_ void* Context, _Inout_ QUIC_STREAM_EVENT* Event) {
+        auto TestContext = (EcnTestContext*)Context;
+        if (Event->Type == QUIC_STREAM_EVENT_RECEIVE) {
+            TestContext->ServerStreamRecv.Set();
+            return QUIC_STATUS_PENDING;
```
Comment (nibanks) [inline src/test/lib/DataTest.cpp:2174]: Why async receives?

File: src/core/settings.c [hunk 16]
```diff
@@ -1188,6 +1213,7 @@ QuicSettingsDump(
     QuicTraceLogVerbose(SettingCongestionControlAlgorithm,  "[sett] CongestionControlAlgorithm = %hu", Settings->CongestionControlAlgorithm);
     QuicTraceLogVerbose(SettingDestCidUpdateIdleTimeoutMs,  "[sett] DestCidUpdateIdleTimeoutMs = %u", Settings->DestCidUpdateIdleTimeoutMs);
     QuicTraceLogVerbose(SettingGreaseQuicBitEnabled,        "[sett] GreaseQuicBitEnabled   = %hhu", Settings->GreaseQuicBitEnabled);
+    QuicTraceLogVerbose(SettingEcnEnabled,                  "[sett] EcnEnabled   = %hhu", Settings->EcnEnabled);
```
Comment (nibanks) [inline src/core/settings.c:1216]: nit: align the `= %hhu` part with the rest above.

File: src/core/connection.h [hunk 17]
```diff
@@ -563,6 +564,18 @@ typedef struct QUIC_CONNECTION {
     //
     QUIC_LOSS_DETECTION LossDetection;
 
+    //
+    // ECT and CE counters for different packet number spaces.
+    //
+    uint64_t EcnEctCounters[QUIC_ENCRYPT_LEVEL_COUNT];
+    uint64_t EcnCeCounters[QUIC_ENCRYPT_LEVEL_COUNT]; // maps to ecn_ce_counters in RFC 9002.
```
[the remark this answers is not in the document]
Comment (csujedihy) [reply src/core/connection.h:571]: Yes, that would be cleaner.

File: src/core/path.h (hunk 5 above)
Comment (csujedihy) [reply src/core/path.h:112]: I thought about this but I don't know if this is the convention, using a possible value as sentinel, in QUIC.

File: src/core/packet_builder.c (hunk 10 above)
Comment (csujedihy) [reply src/core/packet_builder.c:278]: If we move it `QuicSendFlush` or `QuicPacketBuilderInitialize`, which is essentially delaying the ECN decision, we will have to update each sent packet's meta data in `QuicSendFlush`, which will be an extra loop. It also conflicts with `CxPlatSendDataAlloc` where it implies the decision should be made when allocating the send context.

Comment (nibanks) [reply src/core/packet_builder.c:278]: `QuicSendFlush` or `QuicPacketBuilderInitialize` is before this code path here. I'm suggesting you take the logic you have here, calculate what the ECN info would be for the packets, cache it on the Builder, and then use that value here instead.

File: src/core/loss_detection.c (hunk 7 above)
Comment (csujedihy) [reply src/core/loss_detection.c:1512]: It is a real possibility. That's why we need to validate the counts.

File: src/core/packet_builder.c (hunk 10 above)
Comment (csujedihy) [reply src/core/packet_builder.c]: Done

File: src/test/lib/DataTest.cpp (hunk 15 above)
Comment (csujedihy) [reply src/test/lib/DataTest.cpp]: Fixed

File: src/core/loss_detection.c (hunk 8 above)
Comment (csujedihy) [reply src/core/loss_detection.c:1531]: This check is actually necessary. Otherwise, we would emit an ECN validation success log every time we receive a correctly marked packet.

File: src/core/loss_detection.c [hunk 18]
```diff
@@ -1496,6 +1498,69 @@ QuicLossDetectionProcessAckBlocks(
     }
 
     if (NewLargestAck) {
+        if (Connection->Settings.EcnEnabled && Path->EcnValidationState != ECN_VALIDATION_FAILED) {
```
[the remark this answers is not in the document]
Comment (csujedihy) [reply src/core/loss_detection.c:1501]: We don't.

File: src/core/connection.c [hunk 19]
```diff
@@ -5940,6 +5941,16 @@ QuicConnResetIdleTimeout(
         IdleTimeoutMs = Connection->Settings.HandshakeIdleTimeoutMs;
     }
 
+    if (Path->EcnValidationState == ECN_VALIDATION_TESTING) {
+        uint64_t Now = CxPlatTimeUs64();
+        if (CxPlatTimeAtOrBefore64(Now, Path->EcnTestingEndingTime)) {
+            //
+            // Give ECN validation enough time to test ECN capability on the path.
+            //
+            IdleTimeoutMs += US_TO_MS(CxPlatTimeDiff64(Now, Path->EcnTestingEndingTime));
```
Comment (nibanks) [inline src/core/connection.c:5950]: Why would you extend the idle time just to keep testing the network? If they want to kill the connection sooner, we should let them, right?

File: src/core/connection.c [hunk 20]
```diff
@@ -7154,6 +7166,11 @@ QuicConnApplyNewSettings(
             Connection->State.FixedBit = (RandomValue % 2);
             Connection->Stats.GreaseBitNegotiated = TRUE;
         }
+
+        if (Connection->Settings.EcnEnabled) {
+            QUIC_PATH* Path = &Connection->Paths[0];
+            Path->EcnValidationState = ECN_VALIDATION_TESTING;
```
Comment (nibanks) [inline src/core/connection.c:7172]: Should we have ETW logs for every time a path changes ECN validation state? It shouldn't happen that often, right?

File: src/core/connection.c (hunk 19 above)
Comment (csujedihy) [reply src/core/connection.c:5950]: Because we can't tell if the connection is idle or all packets sent in ECN testing state are dropped. We have a similar idle timeout extending logic right below this block.

File: src/core/connection.c (hunk 20 above)
Comment (csujedihy) [reply src/core/connection.c:7172]: I can add one for transiting to unknown state. We have logs for capable and failed.

Comment (nibanks) [reply src/core/connection.c:7172]: I would like a single (manifested) `PathEcnStateChange` event that we could automate processing in WPA if we want, please.

Comment (csujedihy) [reply src/core/connection.c:7172]: What's the point of automating this in WPA? In most cases, the validation succeeds right after the initial packet.

Comment (csujedihy) [reply src/core/connection.c:7172]: When there's actually something interesting happening, the state transition isn't quite useful. What really useful are the counter values in EcnValidationFailure log.

File: src/core/connection.c (hunk 19 above)
Comment (nibanks) [reply src/core/connection.c:5950]: If we're actively sending packets, then we aren't idle. I don't think this change should be necessary.

File: src/core/connection.h [hunk 21]
```diff
@@ -563,6 +563,12 @@ typedef struct QUIC_CONNECTION {
     //
     QUIC_LOSS_DETECTION LossDetection;
 
+    //
+    // The total number of packets sent with each corresponding ECT codepoint in all encryption
+    // level.
+    //
+    uint64_t NumPacketsSentWithEct;
```
Comment (nibanks) [inline src/core/connection.h:570]: I think this kind of info should be put in `QUIC_SEND`.

File: src/core/loss_detection.c (hunk 8 above)
Comment (nibanks) [reply src/core/loss_detection.c:1530]: You resolve this, but didn't remove the check?

File: src/core/loss_detection.c (hunk 9 above)
Comment (nibanks) [reply src/core/loss_detection.c:1809]: We should discuss this in our meeting today.

File: src/core/send.c [hunk 22]
```diff
@@ -1150,6 +1150,30 @@ QuicSendFlush(
     }
     _Analysis_assume_(Builder.Metadata != NULL);
 
+    if (Builder.Path->EcnValidationState == ECN_VALIDATION_CAPABLE) {
+        Builder.EcnEctSet = TRUE;
+    } else if (Builder.Path->EcnValidationState == ECN_VALIDATION_TESTING) {
+        if (Builder.Path->EcnTestingEndingTimeSet) {
+            if (!CxPlatTimeAtOrBefore64(
+                    CxPlatTimeUs64(), Builder.Path->EcnTestingEndingTime)) {
```
Comment (nibanks) [inline src/core/send.c:1158]:
```suggestion
                    TimeNow, Builder.Path->EcnTestingEndingTime)) {
```
We've already paid the cost to query time above. Let's use that.

File: src/core/path.h (hunk 5 above)
Comment (nibanks) [reply src/core/path.h:112]: Let's just use `0`. For 64-bit time values zero is never valid.

File: src/core/send.c [hunk 23]
```diff
@@ -1150,6 +1150,30 @@ QuicSendFlush(
...
+                Builder.Path->EcnValidationState = ECN_VALIDATION_UNKNOWN;
+                QuicTraceLogConnInfo(
+                    EcnValidationUnknown,
+                    Connection,
+                    "ECN unknown.");
+            }
+        } else {
+            uint32_t ThreePtosInUs =
+                QuicLossDetectionComputeProbeTimeout(
```
Comment (nibanks) [inline src/core/send.c:1167]: This will always use the default initial RTT for an estimate. Is that ok?

File: src/core/sent_packet_metadata.h (hunk 2 above)
Comment (nibanks) [reply src/core/sent_packet_metadata.h:97]: Did you validate on Release or Debug?

Comment (csujedihy) [reply src/core/sent_packet_metadata.h:97]: I validated it on both. It's 80B regardless of build flavor.

File: src/core/path.h (hunk 5 above)
Comment (csujedihy) [reply src/core/path.h]: updated

File: src/core/connection.c (hunk 19 above)
Comment (csujedihy) [reply src/core/connection.c]: removed

File: src/core/connection.h (hunk 21 above)
Comment (csujedihy) [reply src/core/connection.h]: done

File: src/core/send.c (hunk 23 above)
Comment (csujedihy) [reply src/core/send.c:1166]: `ThreePtosInUs` for a new connection is 3s. I can cap it if you want.

File: src/core/loss_detection.c [hunk 24]
```diff
@@ -1745,7 +1806,8 @@ QuicLossDetectionProcessTimerOperation(
         // OldestPacket has been in the SentPackets list for at least
         // DisconnectTimeoutUs without an ACK for either OldestPacket or for any
         // packets sent more than the reordering threshold after it. Assume the
-        // path is dead and close the connection.
+        // path is dead and close the connection unless we are validating ECN
+        // capability on the path.
```
Comment (nibanks) [inline src/core/loss_detection.c:1810]: revert this comment change?

File: src/core/loss_detection.c [hunk 25]
```diff
@@ -1710,7 +1772,6 @@ QuicLossDetectionProcessTimerOperation(
     )
 {
     QUIC_CONNECTION* Connection = QuicLossDetectionGetConnection(LossDetection);
-
```
Comment (nibanks) [inline src/core/loss_detection.c:1713]: nit: unnecessary change?

File: src/core/send.h [hunk 26]
```diff
@@ -304,6 +304,12 @@ typedef struct QUIC_SEND {
     //
     uint16_t InitialTokenLength;
 
+    //
+    // The total number of packets sent with each corresponding ECT codepoint in all encryption
+    // level.
+    //
+    uint64_t NumPacketsSentWithEct;
```
Comment (nibanks) [inline src/core/send.h:311]: For slightly better padding, put this after `LastFlushTime`?

File: src/core/connection.c (hunk 20 above)
Comment (nibanks) [reply src/core/connection.c:7178]: As discussed over IM, should anything be promoted to ETW?

Comment (csujedihy) [reply src/core/connection.c:7183]: Added connection rundown ETW event and ecn failure ETW event.

File: src/core/loss_detection.c [hunk 27]
```diff
@@ -1496,6 +1498,66 @@ QuicLossDetectionProcessAckBlocks(
...
+                    EcnValidated = FALSE;
+                } else {
+                    //
+                    // TODO: Notify CC of the ECN signal before we update the CE counts.
+                    //
+                    Packets->EcnCeCounter = Ecn->CE_Count;
+                    Packets->EcnEctCounter = Ecn->ECT_0_Count;
+                    if (Path->EcnValidationState <= ECN_VALIDATION_UNKNOWN) {
+                        Path->EcnValidationState = ECN_VALIDATION_CAPABLE;
+                        QuicTraceLogConnInfo(
```
Comment (nibanks) [inline src/core/loss_detection.c:1530]: Shouldn't this be the same `ConnEcnCapable` event as in the rundown?

File: src/tools/etw/quicetw.h [hunk 28]
```diff
@@ -344,6 +344,7 @@ typedef enum QUIC_EVENT_ID_CONNECTION {
     EventId_QuicConnClientReceivedVersionList,
     EventId_QuicConnServerSupportedVersionList,
     EventId_QuicConnBbr,
+    EventId_QuicConnEcnCapable,
```
Comment (nibanks) [inline src/tools/etw/quicetw.h:347]: EcnFailed event too?

Comment (csujedihy) [reply src/tools/etw/quicetw.h:347]: added

File: src/core/loss_detection.c (hunk 27 above)
Comment (csujedihy) [reply src/core/loss_detection.c]: fixed

**Review by nibanks — APPROVED** [review]
